ratelimits/limiter.go

```diff
@@ -265,61 +251,62 @@ func (l *Limiter) Refund(ctx context.Context, name Name, id string, cost int64)
 	// Remove cancellation from the request context so that transactions are not
 	// interrupted by a client disconnect.
 	ctx = context.WithoutCancel(ctx)
-	tat, err := l.source.Get(ctx, bucketKey(name, id))
+	tat, err := l.source.Get(ctx, bucket.key)
 	if err != nil {
 		return nil, err
 	}
-	d := maybeRefund(l.clk, limit, tat, cost)
+	d := maybeRefund(l.clk, limit, tat, bucket.cost)
 	if !d.Allowed {
 		// The bucket is already at maximum capacity.
 		return d, nil
 	}
-	return d, l.source.Set(ctx, bucketKey(name, id), d.newTAT)
-
+	return d, l.source.Set(ctx, bucket.key, d.newTAT)
 }
 
-// Reset resets the specified bucket.
-func (l *Limiter) Reset(ctx context.Context, name Name, id string) error {
+// Reset resets the specified bucket to its maximum capacity. The new bucket
+// state is persisted to the underlying datastore before returning.
+func (l *Limiter) Reset(ctx context.Context, bucket Bucket) error {
 	// Remove cancellation from the request context so that transactions are not
 	// interrupted by a client disconnect.
 	ctx = context.WithoutCancel(ctx)
-	return l.source.Delete(ctx, bucketKey(name, id))
+	return l.source.Delete(ctx, bucket.key)
 }
 
-// initialize creates a new bucket, specified by limit name and id, with the
-// cost of the request factored into the initial state.
-func (l *Limiter) initialize(ctx context.Context, rl limit, name Name, id string, cost int64) (*Decision, error) {
-	d := maybeSpend(l.clk, rl, l.clk.Now(), cost)
+// initialize creates a new bucket and sets its TAT to now, which is equivalent
+// to a full bucket. The new bucket state is persisted to the underlying
+// datastore before returning.
+func (l *Limiter) initialize(ctx context.Context, rl limit, bucket BucketWithCost) (*Decision, error) {
+	d := maybeSpend(l.clk, rl, l.clk.Now(), bucket.cost)
 
 	// Remove cancellation from the request context so that transactions are not
 	// interrupted by a client disconnect.
 	ctx = context.WithoutCancel(ctx)
-	err := l.source.Set(ctx, bucketKey(name, id), d.newTAT)
+	err := l.source.Set(ctx, bucket.key, d.newTAT)
 	if err != nil {
 		return nil, err
 	}
 	return d, nil
 
 }
 
-// GetLimit returns the limit for the specified by name and id, name is
-// required, id is optional. If id is left unspecified, the default limit for
-// the limit specified by name is returned. If no default limit exists for the
-// specified name, ErrLimitDisabled is returned.
-func (l *Limiter) getLimit(name Name, id string) (limit, error) {
+// GetLimit returns the limit for the specified by name and bucketKey, name is
+// required, bucketKey is optional. If bucketKey is left unspecified, the
+// default limit for the limit specified by name is returned. If no default
+// limit exists for the specified name, errLimitDisabled is returned.
+func (l *Limiter) getLimit(name Name, bucketKey string) (limit, error) {
```

Comment:
What if we use the `bucket` as the only arg and pick through the bucket inside the function? It makes the task slightly simpler at each call site.

Reply:
I actually _almost_ took this approach myself! However, the idea of asking a caller to construct a bucket where `key` is optional felt kinda weird.